### DISCUSSION TITLE
Add "follow user heading" state to map

### DISCUSSION THREAD

**Discussion by okdistribute** [digidem/mapeo-mobile#790] [Ideas]
"Is it possible to make the map rotate as you move to help with navigation to points you want to visit again. Has been using it a lot in the RCA, but this would help, to at least know what direction you are moving in."

Requested by David

**Comment by gmaclennan** [comment]
I think the best way to implement this is by adding a 3rd state to the "locate" button:

<img width="465" alt="image" src="https://user-images.githubusercontent.com/290457/82200619-8691b580-98ff-11ea-9d5e-1e4982321c8d.png">

Currently there are two states:

1. "Map is controlled by user swiping / zooming screen"
2. "Make the map follow user location"

Currently interacting with the map will put the map into state (1). Pressing the locate button in the bottom-right will put the map back into state 2.

In Google Maps and Apple Maps pressing the "Locate" button when it is in state (2) (follow-mode) will put the map into "follow heading" mode, e.g. the map will rotate to follow the user heading.

Mapbox supports this mode in the [`<Camera />` component](https://github.com/react-native-mapbox-gl/maps/blob/master/docs/Camera.md). It has two options:

1. Follow user heading, which uses the magnetometer (compass) in the phone e.g. it will use the phone orientation to rotate the map.
2. Follow user course, which uses the direction of travel to rotate the map.

Option 2 (follow course) is more reliable when the user is moving fast (e.g. in a car) because the magnetometer is subject to other magnetic fields and calibration issues. However, when walking and with poor GPS signal, Option 2 (follow course) is likely to be unreliable, so Option 1 is probably a better option for our case. I _think_ that Apple and Google maps automatically move between heading and course based on reading the phone movement sensors. This might be an option for us in the future.

Finally, currently we do not use the native "follow mode" from Mapbox maps, in order to reduce battery usage, because it resulted in frequent redraws of the map which was using lots of CPU. Instead when in the follow location mode the map updates every time the GPS displaces by 10m or more. We can keep this behaviour, and switch to the native map heading mode only when the user switches into follow heading state.
